docs/dep/dep-004.md

```diff
@@ -0,0 +1,146 @@
+since it can be easily uploaded and distributed to an OCI registry.
+
+The second stage begins as soon as the `draconctl` is invoked to deploy a
+Pipeline. The Pipeline lists a set of components. The user has 3 different
+formats to describe a component:
+
+1. a local filesystem path
+
+This is used by the system to, on the fly read a Task and deploy it on the
+cluster. There are not going to be any checks if the Task already exists, or if
+the version of the Task is newer than the Task being fetched. It's going to be
+`applied` on the cluster, as if the user is invoking `kubectl apply ...`.
+
+2. a URL
+
+The URL is going to be resolved and the Task is going to be fetched and
+deployed on the cluster similar to the case where we have a file in the local
+filesystem.
+
+3. a pURL-inspired id of a Task deployed using Helm
+
+As mentioned previously, we wish to package all our Tasks in one Helm package
+that can be easily deployed on a cluster and provides access to all our
+components. When a user wants to reference a Task deployed via Helm on the
+cluster in a Pipeline description, they will have to use a new URI type that is
+inspired by `pURL` and looks like this: `helm://{{ helm_release_name }}/name`.
+
+The `helm_release_name` is the name of the release chosen by the operator when
+deploying the package, so it's specific to the deployment context. The name of
+the Task will be the actual name of the object. The system will discover all
+the Tasks belonging to the Helm package by filtering with the label
+`app.kubernetes.io/managed-by: Helm` and then the label
+`app.kubernetes.io/instance: { helm_release_name }` and will expect to find the
+Task. If the Task is not found, an error will be returned.
+
+## Implementation
+
+A new interface will be added to the system that can be used for an actual
+implementation but also for mocks for testing purposes:
+
+```golang
+type Applier struct {
+    Apply()
+}
+
+type Component struct {
+    ID() string
+
+    TektonV1Beta1Task() (*tektonv1beta1api.Task, error)
+
+    ManagedByHelm() bool
+}
+
+type Orchestrator interface {
+    Prepare([]Component) error
+
+    Deploy(Applier) error
+}
+```
+
+The `Prepare` method takes as an argument the list of Tasks that are part of
+the Pipeline and ensures that they are deployed. If the Tasks are managed by
+Helm it will make sure that they are available, if they are local or remote
+files, they will be applied on the cluster. If no error is returned the client
+can invoke the `Deploy` method which will deploy the Pipeline itself on the
+cluster.
+
+It is important to note, that the Orchestrator does not manage the assembling
+of a Pipeline or the validation. It expects to receive a correct object and
+will try to deploy whatever pipeline isprovided.
+
+The `Component` struct is an abstraction layer on top of any component we are
+using. It will internally resolve the Task URI and figure out how to manage it,
```

Comment:
> It will internally resolve the Task URI and figure out how to manage it,
If we make everything be a HELM task, then we don't need multiple managers.
Local filesystem management is only useful for development and I'd say, maybe even relatively useless since we can build and load charts easily on local registries.

Furthermore, having the ability to treat everything as a "remote" chart, allows us  to move away from Tekton task defs and support a number of execution backends if we want to

Reply:
I don't see how Helm has anything to with which backend we are using. we could be using Helm to package Tekton V1Beta1 or V1 Tasks, custom deployments and CRDs or whatever else we want. Helm doesn't know or care about the type of object, it just knows that it's some YAML that potentially the K8s API master will recognise with probably some Go templating in it. For me the question of allowing extra ways of fetching a Task is more about development experience as I mentioned in a previous comment

Reply:
sure, makes sense, then please disregard my comment